# Homework 3 Jupyter Notebook
## Sample Python Program

In the following cell, we include a python program that connects to the `class_db` MySQL database in our container and prints the tuples in the Student and Class tables. We will explain the code step-by-step soon, but please execute the code in the cell first and look at the printed tuples:

In [1]:
import mysql.connector

connection = mysql.connector.connect(
  host="localhost",
  user="msa402",      # your username
  password="",        # your pasword
  database="class_db" # your database name
)
connection.autocommit = True

print("Student Table")
print("-------------")
cursor = connection.cursor()
cursor.execute("select * from Student;")
for row in cursor:
    print(row)
cursor.close()

print("\nClass Table")
print("-----------")
cursor = connection.cursor()
cursor.execute("select * from Class;")
for row in cursor:
    print(row)
cursor.close()

connection.close()

Student Table
-------------
(128, 'Peltier', 'Comp. Sci.', 102)
(12345, 'Shankar', 'Comp. Sci.', 32)
(19991, 'Brandt', 'History', 80)
(23121, 'Chavez', 'Finance', 110)
(44553, 'Zhang', 'Physics', 56)
(45678, 'Levy', 'Physics', 46)
(54321, 'Williams', 'Comp. Sci.', 54)
(55739, 'Sanchez', 'Music', 38)
(70557, 'Snow', 'Physics', 0)
(76543, 'Brown', 'Comp. Sci.', 58)
(76653, 'Aoi', 'Elec. Eng.', 60)
(98765, 'Zhang', 'Elec. Eng.', 98)
(98988, 'Tanaka', 'Biology', 120)

Class Table
-----------
('BIO-101', 'Intro. to Biology', 'Biology', 4)
('BIO-301', 'Genetics', 'Biology', 4)
('BIO-399', 'Computational Biology', 'Biology', 3)
('CS-101', 'Intro. to Computer Science', 'Comp. Sci.', 4)
('CS-190', 'Game Design', 'Comp. Sci.', 4)
('CS-315', 'Robotics', 'Comp. Sci.', 3)
('CS-319', 'Image Processing', 'Comp. Sci.', 3)
('CS-347', 'Database System Concepts', 'Comp. Sci.', 3)
('EE-181', 'Intro. to Digital Systems', 'Elec. Eng.', 3)
('FIN-201', 'Investment Banking', 'Finance', 3)
('HIS-351', 'World Hi

### Explanation of the Code

Here is a line-by-line explanation of the above code:

Python needs a special driver to access the MySQL database. In our container, we have already installed the "MySQL-Connector" driver, which can be used by "importing" it:

```python
import mysql.connector
```

Once the driver is imported, you can connect to the MySQL database server using the driver's `connect()` function:

```python
connection = mysql.connector.connect(
  host="localhost",
  user="msa402",      # your username
  password="",        # your pasword
  database="class_db" # your database name
)
```

The next line in the code enables the "autocommit mode" of the driver:

```python
connection.autocommit = True. 
```

When the autocommit mode is on, the driver automatically call the `commit()` function after every SQL statement, so that the effect of the SQL statement is permanetly "committed" in the database. When the autocommit mode is off (the default setting), you will have to maually call the `commit()` function after every SQL statement that may change the database, but this is too easy to forget and inconvenient for doing our homework. 

We are now ready to issue a SQL statement to the MySQL database using the established connection. In order to issue a SQL command, obtain a "cursor" from the connection and call `execute("Your SQL command")` on the cursor:

```python
cursor = connection.cursor()
cursor.execute("select * from Student;")
```

The result from the executed command can be accessed by iterating over the cursor:

```python
for row in cursor:
    print(row)
```

When you finish accessing the result, make sure to "close" the cursor, so that the resources associated with the cursor are freed by the driver:

```python
cursor.close()
```

If you want to run another query, obtain a new cursor from the connection and repeat the process:

```python
cursor = connection.cursor()
cursor.execute("select * from Class;")
for row in cursor:
    print(row)
cursor.close()
```

After you are done with interacting with the database and do not need the database connection any more, make sure to close the connection. Leaving unused open connections can lead to several unexpected errors and performance issues.

```python
connection.close()
```

## Writing Advanced SQL Queries

Now that you know how to access MySQL database with Python, your next job is to write SQL queries given our English descriptions.

To minimize your typing and to help our autograder run your code easily, we define a function named `query()` in the next cell. The function takes a SQL command as an input parameter and prints its result tuples. Execute the next cell so that the `query()` function is available in the Jupyter notebook.

In [2]:
import mysql.connector

def query(sql):
    """Takes a SQL query and prints the resulting tuples from the query"""
    global connection
    if "connection" not in globals() or not connection.is_connected():
        # open a new connection if it is not available
        connection = mysql.connector.connect(host="localhost", user="msa402", password="", database="class_db")
        connection.autocommit = True
    
    cursor = connection.cursor()
    cursor.execute(sql)
    for row in cursor:
        print(row)
    cursor.close()

Now you job is to write SQL queries based on their English descriptions given later. Note that all your answers must be **a single Python statement that calls the `query()` function with a SQL statement** that computes the correct result. For example, if our question is "Return the ID of the instructor 'Srinivasan'", your answer should be

```python
query("select id from Instructor where name='Srinivasan';")
```

Note that in Python, a string can span over multiple lines using the triple-quotes notation:

```python
# multiple-line string
"""
select id
from Instructor
where name='Srinivasan';
"""
```

This notation can be convenient if you need to write a long SQL query that should span over multiple lines.


## Problems

In writing your queries, you may assume that 

-  Every department offers at least one class.
-  Every department has at least one instructor.
-  Every department has at least one student.

Please make sure that your queries **do not return any duplicates,** but try to avoid using `DISTINCT` when it is not necessary
**Query1**: Find the instructors' average salary.

In [3]:
query(
    """ 
    SELECT AVG(salary)
    FROM Instructor;
    """
)



(Decimal('74833.333333'),)


**Query2**: For each department, find the max credit among all classes offered by the department. Your answer should consists of tuples with two columns (dept, maximum_course_credit).

In [4]:
query(
    """
    SELECT dept, MAX(credits)
    FROM Class
    GROUP BY dept
    """
)



('Biology', 4)
('Comp. Sci.', 4)
('Elec. Eng.', 3)
('Finance', 3)
('History', 3)
('Music', 3)
('Physics', 4)


**Query3**: Return the department names that offer at least three classes.

In [5]:
query(
    """
    SELECT dept
    FROM Class
    GROUP BY dept
    HAVING COUNT(*) >= 3;
    """
)



('Biology',)
('Comp. Sci.',)


**Query4**: Find the names of departments that offer only 3-credit classes if they offer any class at all. Include a department in your answer if it offers no class.

In [6]:
query(
    """
    SELECT dept
    FROM Class
    WHERE dept NOT IN
    
    (SELECT dept
    FROM Class
    WHERE credits = 4);
    """
)




('Elec. Eng.',)
('Finance',)
('History',)
('Music',)


**Query5**: Return the average credit of the courses that are offered by the 'Comp. Sci.' department. 

In [7]:
query(
    """
    SELECT AVG(credits)
    FROM Class
    WHERE dept = 'Comp. Sci.';
    """
)



(Decimal('3.4000'),)


**Query6**: As above, but display the average course credit per *every* department. Your answer should consist of two columns (dept, dept_average_course_credit) and every department should appear once in your result.

In [8]:
query(
    """
    SELECT dept, AVG(credits)
    FROM Class
    GROUP BY dept;
    """
)


('Biology', Decimal('3.6667'))
('Comp. Sci.', Decimal('3.4000'))
('Elec. Eng.', Decimal('3.0000'))
('Finance', Decimal('3.0000'))
('History', Decimal('3.0000'))
('Music', Decimal('3.0000'))
('Physics', Decimal('4.0000'))


**Query7**: For every department, display the average course credit of the department **and** the *overall average* of the course credits (i.e., the average credit over *all* courses, not within a department). Your answer should consist of three columns (dept, dept_avg_course_credit, overall_avg_course_credit). In your answer, return one tuple per each department.

Note that the overall course-credit average is 3.3846 in our dataset, so something is wrong with your query if you get a different number in the third column of your result.

In [9]:
query(
    """
    SELECT DISTINCT dept, AVG(credits) OVER(PARTITION BY dept), AVG(credits) OVER()
    FROM Class;
    """
)



('Biology', Decimal('3.6667'), Decimal('3.3846'))
('Comp. Sci.', Decimal('3.4000'), Decimal('3.3846'))
('Elec. Eng.', Decimal('3.0000'), Decimal('3.3846'))
('Finance', Decimal('3.0000'), Decimal('3.3846'))
('History', Decimal('3.0000'), Decimal('3.3846'))
('Music', Decimal('3.0000'), Decimal('3.3846'))
('Physics', Decimal('4.0000'), Decimal('3.3846'))


**Query8**: Compute the average salary of all instructors and compare it against the average instructor salary within each department. Your answer must consists of two columns (dept, diff_avg_salary), where dept is a department name and diff_avg_salary is avg(all instructor salaries) - avg(the department's instructors' salaries). In your answer, return one tuple per each department.

In [10]:
query(
    """
    SELECT distinct dept, (AVG(salary) OVER() - AVG(salary) OVER(PARTITION BY dept) )
    FROM Instructor;
    """
)



('Biology', Decimal('2833.333333'))
('Comp. Sci.', Decimal('-2500.000000'))
('Elec. Eng.', Decimal('-5166.666667'))
('Finance', Decimal('-10166.666667'))
('History', Decimal('13833.333333'))
('Music', Decimal('34833.333333'))
('Physics', Decimal('-16166.666667'))


**Query9**: 'Comp. Sci.' and 'Elec. Eng.' departments belong to 'Engineering' school and all other departments belong to 'L&S' (letter and science) school. For every department, return the school they belong to. Your answer must consist of two columns (dept, school). 

Remark: There are many ways to write this query, but using the `CASE` function is likely to lead to the most succinct query.

In [11]:
query(
    """
    SELECT dept, CASE WHEN(dept='Comp. Sci.' OR dept='Elec. Eng.') THEN 'Engineering' ELSE 'L&S' END AS school
    FROM Department;
    """
)


('Biology', 'L&S')
('Comp. Sci.', 'Engineering')
('Elec. Eng.', 'Engineering')
('Finance', 'L&S')
('History', 'L&S')
('Music', 'L&S')
('Physics', 'L&S')


**Query10**: For each school (i.e., Engineering or L&S), show the number of students and the number of instructors in the school. Your answer must consists of three columns (school, num_studs, num_insts). 


In [12]:
query(
    """
    SELECT CASE WHEN(D.dept='Comp. Sci.' OR D.dept='Elec. Eng.') THEN 'Engineering' ELSE 'L&S' END AS school, COUNT(DISTINCT S.name) AS num_studs, COUNT(DISTINCT I.name) AS num_insts
    FROM Department D LEFT OUTER JOIN Student S ON D.dept = S.dept LEFT OUTER JOIN Instructor I ON D.dept = I.dept
    GROUP BY school;
    """
)


('Engineering', 6, 4)
('L&S', 7, 8)


**Query11**: Find the id's of the students who took more classes in 2009 than in 2010 (in terms of number of classes, not credits).
Remark: This query is quite tricky to get right because some students took classes in 2009 but not in 2010. Using `OUTER JOIN` may lead to a succinct query.


In [13]:
query(
    """
    WITH Classes AS 
    (SELECT stud_id, COUNT(CASE WHEN(year='2009') THEN '1' ELSE NULL END) AS Class_09, COUNT(CASE WHEN(year='2010') THEN '1' ELSE NULL END) AS Class_10
    FROM Takes
    GROUP BY stud_id
    HAVING Class_09 > Class_10)
    SELECT stud_id FROM Classes;
    """
)


(128,)
(12345,)
(44553,)
(54321,)
(76653,)


**Query12**: For every student, return the year in which they obtained the maximum number of course credits. Your answer must consists of two columns, (stud_id, max_credit_year). If a student, say id 12345, obtained the maximum number of course credits in multiple years, say in 2009 and 2010, your answer must return one tuple per each year, like (12345, 2009) and (12345, 2010). If a student did not take any class at all, do not return the student.

Remark: There are many ways to write this query, but using *common table expression* may lead to a succinct query.

In [14]:
query(
    """
    WITH MaxCred AS
    (SELECT stud_id, year, SUM(credits) AS tot_creds, MAX(SUM(credits)) OVER(PARTITION BY stud_id) AS highest
    FROM Takes
    LEFT JOIN Class ON Takes.class_id = Class.id
    GROUP BY stud_id, year)
    
    SELECT stud_id, year
    FROM MaxCred
    WHERE highest = tot_creds;
    """
)



(128, 2009)
(12345, 2009)
(19991, 2010)
(23121, 2010)
(44553, 2009)
(45678, 2010)
(54321, 2009)
(55739, 2010)
(76543, 2009)
(76653, 2009)
(98765, 2009)
(98988, 2009)
(98988, 2010)


**Query13**: Return the id's of the top-4 students who have obtained the largest number of course credits so far. In case of a tie, you can break the tie arbitrarily (i.e., you may return any student(s) that tied). 

Remark: MySQL does not support the SQL standard `FETCH FIRST` clause. You may have to use the `LIMIT` clause that is supported by MySQL for this query.

In [30]:
query(
    """
    WITH TotCred AS
    (SELECT stud_id, SUM(credits) AS tot_creds
    FROM Takes
    LEFT JOIN Class ON Takes.class_id = Class.id
    GROUP BY stud_id)
    
    SELECT stud_id 
    FROM TotCred
    ORDER BY tot_creds DESC
    LIMIT 4;
    """
)



(12345,)
(45678,)
(98988,)
(54321,)


**Query14**: For every student, show their name and their advisor's name. Your answer must consist of two columns (student_name, advisor_name). If a student does not have any advisor, return NULL as the advisor's name. 

Remark: There are many ways to write this query, but using `OUTER JOIN` may lead to a succinct query.

**Note**: Python does not have NULL as one of its primitive values, so it automatically converts SQL NULL values into `None`. If your answers contain `None` as opposed to NULL, that is expected. No need to worry about it.

In [16]:
query(
    """
    WITH A AS (SELECT Advisor.stud_id, Advisor.inst_id, Instructor.name FROM Instructor, Advisor WHERE Advisor.inst_id = Instructor.id)
    SELECT Student.name, A.name
    FROM Student LEFT OUTER JOIN A ON Student.id = A.stud_id
    GROUP BY Student.id;
    """
)


('Peltier', 'Katz')
('Shankar', 'Srinivasan')
('Brandt', None)
('Chavez', 'Singh')
('Zhang', 'Einstein')
('Levy', 'Einstein')
('Williams', None)
('Sanchez', None)
('Snow', None)
('Brown', 'Katz')
('Aoi', 'Kim')
('Zhang', 'Kim')
('Tanaka', 'Crick')


**Query15**: Some of you may have noticed that the tot\_cred value for students do not match the credits from the courses they have taken. For every student, show this discrepency. Your answer must consist of two columns, (stud_id, credit_discrepency), where credit_discrepency is tot_cred - sum(credits of the courses taken by the student).

Please note that some students may not have taken any classes. Your answer must include those students as well.

Remark: You may find the `COALESCE()` function helpful in writing the query.

In [36]:
query(
    """
    WITH Diff AS 
    (SELECT S.id, COALESCE(SUM(credits), tot_cred) AS real_creds, tot_cred
    FROM Student S
    LEFT JOIN Takes T ON S.id = T.stud_id
    LEFT JOIN Class C ON C.id = T.class_id
    GROUP BY S.id)
    
    SELECT id, (tot_cred - real_creds) AS credit_discrepancy
    FROM Diff;
    """
)



(128, Decimal('95'))
(12345, Decimal('18'))
(19991, Decimal('77'))
(23121, Decimal('107'))
(44553, Decimal('52'))
(45678, Decimal('35'))
(54321, Decimal('46'))
(55739, Decimal('35'))
(70557, Decimal('0'))
(76543, Decimal('51'))
(76653, Decimal('57'))
(98765, Decimal('91'))
(98988, Decimal('112'))


**Query16**: Return all class id's that a student must take before taking 'BIO-399', including its prerequisites and the prerequisites of its prerequisites, etc. Do not include 'BIO-399' in your answer.

Remark: You will need to write a recursive query to compute the answer since we do not know the length of BIO-399's prerequisite chain in advance.

In [18]:
query(
    """
    WITH RECURSIVE Bio_Prereq(class_id, prereq) AS (
    (SELECT class_id, prereq_id AS prereq FROM Prereq)
    UNION
    (SELECT P.class_id, B.prereq
    FROM Prereq P, Bio_Prereq B
    WHERE P.prereq_id = B.class_id)
    )
    SELECT prereq FROM Bio_Prereq WHERE class_id = 'BIO-399';
    """
)



('BIO-301',)
('CS-347',)
('BIO-101',)
('CS-101',)


## What to Submit

Once you finish filling in all cells with your queries, save this notebook and upload it to GradeScope before the deadline. Make sure that the submitted file name is `hw3.ipynb`

Note : The final submission should run without errors. Click on 'Kernel -> Restart and Run All', this should return correct answers in each cell and does not give any errors. Make sure the submission contains only the expected commands. Remove all unncessary code that you added during debugging!

You can submit your work as many times as you want. In case of multiple submissions, the grade from the latest submission will be used. In order to accommodate the last minute snafu during submission, you will have 1-hour window after the deadline to finish your submission process. That is, as long as you start your submission before the deadline and complete within 1 hour after the deadline, everything will be fine.